cpp/src/io/json/nested_json.hpp

```diff
@@ -172,6 +172,54 @@ struct json_column {
                   uint32_t child_count);
 };
 
+/**
+ * @brief Intermediate representation of data from a nested JSON input, in device memory.
+ * Device memory equivalent of `json_column`.
+ */
+struct d_json_column {
+  // Type used to count number of rows
+  using row_offset_t = size_type;
+
+  // The inferred type of this column (list, struct, or value/string column)
+  json_col_t type = json_col_t::Unknown;
+
+  rmm::device_uvector<row_offset_t> string_offsets;
+  rmm::device_uvector<row_offset_t> string_lengths;
+
+  // Row offsets
+  rmm::device_uvector<row_offset_t> child_offsets;
+
+  // Validity bitmap
+  rmm::device_uvector<bitmask_type> validity;
+
+  // Map of child columns, if applicable.
+  // Following "element" as the default child column's name of a list column
+  // Using the struct's field names
+  std::map<std::string, d_json_column> child_columns;
```

Comment:
Since you keep the order in `column_order`:
```suggestion
  std::unordered_map<std::string, d_json_column> child_columns;
```

Reply:
I followed the same types used in `json_column`. will optimize this in a follow up PR.